src/Analyser/MutatingScope.php

```diff
@@ -1774,7 +1774,7 @@ private function resolveType(Expr $node): Type
 						}
 					} else {
 						$arrayBuilder->degradeToGeneralArray();
-						$arrayBuilder->setOffsetValueType(new IntegerType(), $valueType->getIterableValueType());
+						$arrayBuilder->setOffsetValueType(new IntegerType(), $valueType->getIterableValueType(), true);
```

Comment:
There are more situations to cover. For example if you unpack `non-empty-array`, we know the result array is going to be `non-empty-array` too. You can check this with `$valueType->getIterableValueType()->isIterableAtLeastOnce()->yes()`.

Reply:
thank you for the fast input! I hope reviewing this is not more effort than fixing it yourself 😅

Added handling for the non-empty-array cases including some more tests.